keps/sig-network/3698-multi-network-requirements/README.md

```diff
@@ -0,0 +1,375 @@
+networks in the Kubernetes cluster. It would allow attaching a pod to one or
+several networks via a given type of interface depending on its connectivity
+or performance needs.
+
+### Goals
+
+Define user stories and requirements for the Multi-Network effort in Kubernetes.
+
+### Non-Goals
+
+Define API and implementation.
+
+## Proposal
+
+### Personas
+
+#### Network Administrator
+
+With this proposal we will introduce a new “persona”, called **Network
+Administrator**. This role will be responsible for defining and managing
+“Networks” (name TBD, referred as “Object” in the Requirement section) that
+properly describe the infrastructure available for the cluster, namespace and
+workloads. This persona can define which users can “attach” to a specific
+“Network”.
+
+#### User
+
+User is the consumer of Networks (name TBD, referred as “Object” in the
+Requirement section) via referencing them in their workloads. Users usually will
+not create or remove the Network on their own.
+
+### Terminology
+
+* **Default network** - This is the initial cluster-wide Pod networking provided
+  during cluster creation that is available to the Pod when no additional
+  networking configuration is provided.
+* **Primary network** - This is the network inside the Pod which interface is
+  used for the default gateway.
+
+### User Stories
+
+All user stories represent the type of use cases the multi-networking API should
+be able to support. References to technologies or exact products does not
+indicate that this API will directly support them. The set of requirements,
+defined out of these use cases, are the final indicator of what will be covered
+by this API and effort.
+
+#### Story #1
+As a Cloud Native Network Function (CNF) vendor I require an additional
+interface to be provisioned into the Kubernetes Pod. Each of these interfaces
+has to be in an isolated network for regulatory compliance. The isolation has to
+be done on a Layer-2.
+
+<p align="center">
+  <img src="mn-story-1.png?raw=true" alt="multi-network story 1 network L2 isolation"/>
+</p>
+
+#### Story #1a
+As a Cloud NativeNetwork Function (CNF) vendor I require an additional interface
+to be provisioned into the Kubernetes Pod. Each of these interfaces has to be in
+an isolated network for regulatory compliance. The isolation has to be done on a
+Layer-3.
+
+<p align="center">
+  <img src="mn-story-1a.png?raw=true" alt="multi-network story 1a network L3 isolation"/>
+</p>
+
+#### Story #2
+As a Cloud Native Network Function (CNF) vendor I require a HW-based interface
+(e.g. SRIOV VF) to be provisioned to my workload Pod. I need to leverage that HW
+for performance purposes (high bandwidth, low latency), that my user-space
+application (e.g. DPDK-based) can use. The VF will not use the standard netdev
+kernel module. The Pod’s scheduling to the nodes should be based on hardware
+availability (e.g. devicePlugin or some other way).
```

Comment:
Usually, when you dig into this, what you discover is that you have a workload that requires some sort of special treatment from the physical network and a fast path to it.

The problem is that this requirement is entirely mis-focused.   Literally nobody wants an SRIOV VF.  What they want is a fast path to the physical network doing a specific set of things for them.  

You can tell this because 
1. If you offer someone a VF leads to the physical network *mishandling* the packets pushed into the VF, they will be very unhappy.
2. If you can't offer them a SRIOV VF because your hardware only supports SIOV (nexts technology coming after SRIOV) they will be very unhappy (even if you could have given them a vfio for the SIOV that is connected to the proper physical network treatment)
3. If you offer them vfio for the VF when they needed a kernel interface, they will be unhappy.

All of this becomes even more crucial when you are scheduling the Pod.  If I schedule the Pod to a Node that just has SRIOV VFs available, but its NICs lack the required capabilities, or cannot provide a fast path to the required sort of treatment, this all breaks down.

There *are* important sets of requirements in here... but I would recommend thinking them through outside of the 'interface' 'network' paradigm being applied here to get to the real root of the need.

I'd suggest rethinking the requirement in terms of:

1.  What Network Service does the user want from the physical network?
2. How do they want to consume it?  Via what mechanism? (kernel interface, rdma, vfio, etc)
3. What capabilities do they need from the fast path connection to that Network Service?

Reply:
I agree with Dan that this is a particularly compelling user story. As for SRIOV VF I get that this was used as an example and so focusing in on it specifically might be distracting, but I also think it might be worth taking a couple seconds to add a couple more examples just to moot the point.